Service discovery hardening

fcrisciani opened the pull request:
This patch addresses several issues found on the Service Discovery feature.

fcrisciani commented on sandbox.go:
```diff
@@ -669,24 +669,14 @@ func (sb *sandbox) SetKey(basePath string) error {
 func (sb *sandbox) EnableService() error {
 	for _, ep := range sb.getConnectedEndpoints() {
-		if ep.enableService(true) {
-			if err := ep.addServiceInfoToCluster(); err != nil {
```
Restore Enable service logic (introducing again the sb lock)

mavenugo commented on network.go:
```diff
@@ -1886,9 +1891,19 @@ func (n *network) ResolveIP(ip string) string {
+	ipInfo, ok := elemSet[0].(ipInfo)
+	if !ok {
+		setStr, b := sr.ipMap.String(ip)
+		logrus.Warnf("Something is wrong with the ipInfo set for key %s set:%t %s", ip, b, setStr)
```
a better warning message maybe ?

fcrisciani replied:
suggestion? It will be !ok if the element inserted is not of type ipInfo that is strange and major bug

mavenugo commented on network.go:
```diff
@@ -1886,9 +1891,19 @@ func (n *network) ResolveIP(ip string) string {
+	// network db notifications)
+	// In such cases the resolution will be based on the first element of the set, and can vary
+	// during the system stabilitation
+	ipInfo, ok := elemSet[0].(ipInfo)
```
Is the check for `extResolver` intentionally ignored here ?

fcrisciani replied:
nope, will add that also

mavenugo commented on networkdb/networkdb.go:
```diff
@@ -285,7 +285,6 @@ func (nDB *NetworkDB) CreateEntry(tname, nid, key string, value []byte) error {
 	nDB.indexes[byNetwork].Insert(fmt.Sprintf("/%s/%s/%s", nid, tname, key), entry)
 	nDB.Unlock()
 
-	nDB.broadcaster.Write(makeEvent(opCreate, tname, nid, key, value))
```
Can you pls explain why it is okay to remove the broadcast event ?

fcrisciani replied:
This is triggering a notification of the handleEpTableEvent for local endpoints. This is not necessary considering that local endpoints got already configured by logic before. Also that can be risky because now the timing with remote events can arrive out of sync due to the fact that remote events will take more time than local dispatched ones

mavenugo replied:
am not against this change. But I like to understand this a bit more.

`broadcaster.Write` ultimately calls appropriate Watch routines such as `handleEpTableEvent` or `handleNodeTableEvent` or other tables managed by other components (such as overlay driver).

CreateEntry in networkdb is a generic API to create an event of any type. But changing this generic function for a specific case of Endpoint management seems risky without considering how it impacts other events.
What if there is another table (not endpoint), with an event created by one component and consumed not just by remote peers, but also by another local component which networkdb as the mechanism for communication ?

Also, if we remove this call, can you point me to another code path for which handleEpTableEvent is called for local endpoints ?

mavenugo replied:
Based on the review on all the current usage of CreateEntry and Watch, there is no such case that will break and hence am fine with not having to broadcast this event from networkDB.

fcrisciani commented on network.go:
```diff
@@ -1868,9 +1873,9 @@ func (n *network) HandleQueryResp(name string, ip net.IP) {
-		ipInfo.extResolver = true
+	if ok, _ := sr.ipMap.Contains(ipStr, ipInfo{name: name, extResolver: true}); !ok {
+		sr.ipMap.Remove(ipStr, ipInfo{name: name})
+		sr.ipMap.Insert(ipStr, ipInfo{name: name, extResolver: true})
```
If you look at the original logic what the code does is to fetch the element from the map and set the extResolver to true.
The new logic simply checks if there is already an element with extResolver to true, if there is not then removes the entry that has extResolver to false and inserts one to true.

Looking better the logic is slightly different from the original, in the sense that I insert all the time an entry if it is not there, instead the original logic if the entry is not there does not change it. I will add the check to see if the entry is there with extResolver to false

fcrisciani commented on service.go:
```diff
@@ -46,19 +48,51 @@ type service struct {
 
 	// Service aliases
-	aliases []string
+	serviceAliases []string
```
ok will restore

mavenugo commented on network.go:
```diff
@@ -97,7 +98,7 @@ type ipInfo struct {
 	svcMap     map[string][]net.IP
 	svcIPv6Map map[string][]net.IP
-	ipMap      map[string]*ipInfo
+	ipMap      common.SetMatrix
```
I dont quite understand the need to have this ipMap structure to container more than 1 ipInfo ?
Hence I dont see a reason to change it to use SetMatrix.

mavenugo commented on sandbox.go:
```diff
@@ -668,26 +671,25 @@ func (sb *sandbox) SetKey(basePath string) error {
+	logrus.Debugf("DisableService %s START", sb.containerID)
 	for _, ep := range sb.getConnectedEndpoints() {
-		if ep.enableService(false) {
-			if err := ep.deleteServiceInfoFromCluster(); err != nil {
```
Thinking more about this, Isnt it better to have `deleteServiceInfoFromCluster` called from DisableService, but remove it from `sbLeave` in order to be consistent with the way EnableService and DisableService are handled. Also it brings in consistency between Join and Leave. WDYT ?

fcrisciani replied:
That would be great as a future addition, but we have to investigate some corner cases. For example I validated that if you do the `docker kill <container>` no DisableService is called. This of course will leave stale entries behind.

mavenugo replied:
👍

mavenugo commented on service.go:
```diff
@@ -48,17 +50,49 @@ type service struct {
+type loadBalancerBackend struct {
+	ip            net.IP
+	containerName string
+	taskAliases   []string
```
It is weird to see taskAliases and containerName carried in loadBalancerBackend, where this structure is used so far for the VIP based IPVS programming. Are we changing this for other purposes now (such as DNS-RR as well) ?

mavenugo replied:
I can see why we have to carry these additional details. This is required especially so that `cleanupServiceBindings` can call the lower level SD functions that expects to see containerName and taskAliases. But still, this seems out of place.

mavenugo replied:
@fcrisciani Please consider picking up this change https://github.com/mavenugo/libnetwork/commit/92820b94170bc4101b3914859a4ebbcb31c092ca

This is a minor rework of your changes that enables us to avoid the need to change this structure and also helps concentrate on this PR to fixing the issues and minimizing the code-reorg for a later activity if there is a need.

fcrisciani commented:
Don't see this message on github, so will reply by email.

The reasons are several:
1) Events are happening in parallel, imagine this scenario with 2 workers.
Worker A is allocating IPa, now that container exit so IPa is released.
Before that the notification from network db arrives to Worker B, a new
container on Worker B start using IPa. When the notification from WorkerA
arrives to Worker B without the SetMatrix the entry in the ipMap will be
removed. This will break the resolution for the container that is still
running on Worker B.

2) Events from network DB sometime are replayed for endpoints that got
deleted in the past. I saw events delayed up to 6 min. In this case if the
IP is reused by any other worker, the notification arriving will wipe out
the entry from the ipMap.

For all the previous reason, the SetMatrix will guarantee that a specific
IP is going to be removed when no other endpoint is associated with it.

On Sat, Jun 10, 2017 at 1:14 PM, Madhu Venugopal <notifications@github.com>
wrote:

> *@mavenugo* commented on this pull request.
> ------------------------------
>
> In network.go
> <https://github.com/docker/libnetwork/pull/1796#discussion_r121265453>:
>
> > @@ -97,7 +98,7 @@ type ipInfo struct {
>  type svcInfo struct {
>  	svcMap     map[string][]net.IP
>  	svcIPv6Map map[string][]net.IP
> -	ipMap      map[string]*ipInfo
> +	ipMap      common.SetMatrix
>
> I dont quite understand the need to have this ipMap structure to container
> more than 1 ipInfo ?
> Hence I dont see a reason to change it to use SetMatrix.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/libnetwork/pull/1796#pullrequestreview-43305173>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABvTieULfqSMaiyLTPRzIJX7RirJOZrYks5sCvkxgaJpZM4N0i7C>
> .
>

fcrisciani commented on network.go:
```diff
@@ -1886,13 +1893,27 @@ func (n *network) ResolveIP(ip string) string {
+	// In such cases the resolution will be based on the first element of the set, and can vary
+	// during the system stabilitation
+	elem, ok := elemSet[0].(ipInfo)
+	if !ok {
```
the SetMatrix is a generic data structure that store potentially any kind of type. This, as far as I know, is the only way to cast back to the original type.

mavenugo replied:
ok. This is a defensive check and am okay to keep it in and am hoping to never see the Error message ever :)

fcrisciani commented on network.go:
```diff
@@ -1886,13 +1893,27 @@ func (n *network) ResolveIP(ip string) string {
+	elem, ok := elemSet[0].(ipInfo)
+	if !ok {
+		setStr, b := sr.ipMap.String(ip)
+		logrus.Warnf("BUG expected set of ipInfo type for key %s set:%t %s", ip, b, setStr)
```
it has to raise the alarm, It will mean that in that map got inserted a wrong type.

mavenugo commented on agent.go:
```diff
@@ -814,58 +840,44 @@ func (c *controller) handleEpTableEvent(ev events.Event) {
-		n.addSvcRecords(name, ip, nil, true)
-		for _, alias := range taskaliases {
-			n.addSvcRecords(alias, ip, nil, true)
-		}
```
I can see how you have removed this here and consolidated all the addSvcRecords under `addEndpointNameResolution`.

But removing this will result in unmanaged containers in `attachable` networks losing SD.

Hence, we have to call `addEndpointNameResolution` from here and `addEndpointNameResolution` must be changed to also handle unmanaged container scenario.

mavenugo commented on agent.go:
```diff
@@ -814,58 +840,44 @@ func (c *controller) handleEpTableEvent(ev events.Event) {
-		n.deleteSvcRecords(name, ip, nil, true)
-		for _, alias := range taskaliases {
-			n.deleteSvcRecords(alias, ip, nil, true)
-		}
```
Same here.

mavenugo commented on agent.go:
```diff
@@ -602,8 +622,7 @@ func (ep *endpoint) addServiceInfoToCluster() error {
+		if err := c.addServiceBinding(ep.svcName, ep.svcID, n.ID(), ep.ID(), ep.Name(), ep.virtualIP, ingressPorts, ep.svcAliases, ep.myAliases, ep.Iface().Address().IP, "addServiceInfoToCluster"); err != nil {
 			return err
 		}
 	}
```
I think you should also call `addEndpointNameResolution` from here for the case of unmanaged container and its name resolution.

mavenugo commented on agent.go:
```diff
@@ -655,8 +680,7 @@ func (ep *endpoint) deleteServiceInfoFromCluster() error {
+		if err := c.rmServiceBinding(ep.svcName, ep.svcID, n.ID(), ep.ID(), ep.Name(), ep.virtualIP, ingressPorts, ep.svcAliases, ep.myAliases, ep.Iface().Address().IP, "deleteServiceInfoFromCluster"); err != nil {
 			return err
 		}
 	}
```
I think you should also call deleteEndpointNameResolution from here for the case of unmanaged container and its name resolution.

mavenugo commented on service_common.go:
```diff
@@ -6,15 +6,126 @@ import (
+		}
+	}
+
+	if addService && len(vip) != 0 {
```
Its hard for me to judge if `addService` variable is introduced here just as an optimization to avoid calling `addSvcRecords` multiple times for the same svcName <-> vip combination ?
or Is there a real problem if we call `addSvcRecords` multiple times for the same svcName <-> vip combination ?

fcrisciani replied:
no real problem is only an optimization and also more symmetric towards the delSvcRecords that has the rmService (the case of the removal the logic is mandatory of course)

fcrisciani commented on agent.go:
```diff
@@ -593,24 +593,49 @@ func (ep *endpoint) addServiceInfoToCluster() error {
 	c := n.getController()
 	agent := c.getAgent()
 
+	name := ep.Name()
```
@mavenugo I guess to handle properly the case of anonymous container this has to be brought up

fcrisciani commented on agent.go:
```diff
@@ -647,17 +674,33 @@ func (ep *endpoint) deleteServiceInfoFromCluster() error {
-		if n.ingress {
-			ingressPorts = ep.ingressPorts
-		}
+	name := ep.Name()
```
@mavenugo same here for the delete this name is needed
